src/libAtomVM/atom_table.c

```diff
@@ -0,0 +1,474 @@
+}
+
+static inline struct HNode *get_node_from_bucket(
+    const struct AtomTable *hash_table, unsigned long bucket_index, AtomString string)
+{
+    struct HNode *node = hash_table->buckets[bucket_index];
+    while (node) {
+        if (atom_are_equals(string, node->key)) {
+            return node;
+        }
+
+        node = node->next;
+    }
+
+    return NULL;
+}
+
+static inline struct HNode *get_node(const struct AtomTable *hash_table, AtomString string)
+{
+    unsigned long hash = sdbm_hash(string, atom_string_len(string));
+
+    unsigned long bucket_index = hash % hash_table->capacity;
+    return get_node_from_bucket(hash_table, bucket_index, string);
+}
+
+static inline struct HNode *lock_and_get_node(struct AtomTable *hash_table, AtomString string)
+{
+    unsigned long hash = sdbm_hash(string, atom_string_len(string));
+
+    SMP_RDLOCK(hash_table);
+    unsigned long bucket_index = hash % hash_table->capacity;
+    return get_node_from_bucket(hash_table, bucket_index, string);
+}
+
+long atom_table_get_index(struct AtomTable *table, AtomString string)
+{
+    struct HNode *node = lock_and_get_node(table, string);
+    long result = (node != NULL) ? node->index : ATOM_TABLE_NOT_FOUND;
+
+    SMP_UNLOCK(table);
+    return result;
+}
+
+// TODO: this function needs use an efficient structure such as a skip list
+static struct HNode *get_node_using_index(struct AtomTable *table, long index)
+{
+    if (UNLIKELY(index >= table->count)) {
+        return NULL;
+    }
+
+    struct HNodeGroup *node_group = table->first_node_group;
+    while (node_group) {
+        long first_index = node_group->first_index;
+        if (first_index + node_group->len > index) {
+            return &node_group->nodes[index - first_index];
+        }
+
+        node_group = node_group->next;
+    }
+
+    return NULL;
+}
+
+AtomString atom_table_get_atom_string(struct AtomTable *table, long index)
+{
+    SMP_RDLOCK(table);
+
+    struct HNode *node = get_node_using_index(table, index);
+    if (IS_NULL_PTR(node)) {
+        return NULL;
+    }
+
+    AtomString found_key = node->key;
+
+    SMP_UNLOCK(table);
+    return found_key;
+}
+
+int atom_table_cmp_using_atom_index(struct AtomTable *table, int t_atom_index, int other_atom_index)
+{
+    AtomString t_atom_string = atom_table_get_atom_string(table, t_atom_index);
+
+    int t_atom_len = atom_string_len(t_atom_string);
+    const char *t_atom_data = (const char *) atom_string_data(t_atom_string);
+
+    AtomString other_atom_string = atom_table_get_atom_string(table, other_atom_index);
+
+    int other_atom_len = atom_string_len(other_atom_string);
+    const char *other_atom_data = (const char *) atom_string_data(other_atom_string);
+
+    int cmp_size = (t_atom_len > other_atom_len) ? other_atom_len : t_atom_len;
+
+    int memcmp_result = memcmp(t_atom_data, other_atom_data, cmp_size);
+
+    if (memcmp_result == 0) {
+        if (t_atom_len == other_atom_len) {
+            return 0;
```

Comment:
This case should be handled by t_atom_index == other_atom_index, shouldn't it?

Reply:
Yes, right. This branch is useful just to avoid leaking any caller assumption into the callee semantic.